posthog/hogql/resolver.py

```diff
@@ -101,6 +101,12 @@ def visit(self, node: ast.Expr) -> ast.Expr:
         return super().visit(node)
 
     def visit_select_union_query(self, node: ast.SelectUnionQuery):
+        node = cast(ast.SelectUnionQuery, clone_expr(node))
+
+        # All parts of UNION ALL which don't have CTEs can use CTEs from the first statement
+        for select_query in node.select_queries[1:]:
+            select_query.ctes = select_query.ctes or node.select_queries[0].ctes
+
```

Comment:
Cloning is not cheap, and this will clone every select query at least once, no matter how deep. If the only thing you're actually cloning is the CTE-s, then I'd just clone those nodes.

Reply:
My assumption is that resolver clones nodes to keep the initially passed node unmodified. Is that correct? If so, I clone the whole node to copy all select_queries, because I want to modify almost all select_queries by adding ctes. I could copy only the queries that don't have their own ctes, but it seems like a too small optimization.

Maybe my initial assumption is wrong?